samples/nrf9160/serial_lte_modem/src/main.c

```diff
@@ -0,0 +1,289 @@
+		LOG_INF("Sync up");
+	}
+#endif
+#if defined(CONFIG_THIN_GPS_SERVICE)
+	else if (cmd_type == CMD_TYPE_GPS_CMD) {
+		th_gps_control(cmd_type, data_buf);
+	}
+#endif
+#if defined(CONFIG_THIN_SNTP_SERVICE)
+	else if (cmd_type == CMD_TYPE_TIME_CMD) {
+		th_sntp_control(cmd_type);
+	}
+#endif
+#if defined(CONFIG_LTE_LINK_CONTROL)
+	else if ((cmd_type&CMD_TYPE_MDM_BASE) == CMD_TYPE_MDM_BASE) {
+		th_modem_control(cmd_type, data_buf);
+	}
+#endif
+#if defined(CONFIG_THIN_MQTT_SERVICE)
+	else if ((cmd_type&CMD_TYPE_MQTT_BASE) == CMD_TYPE_MQTT_BASE) {
+		th_mqtt_control(cmd_type, data_buf, data_len);
+	}
+#endif
+#if defined(CONFIG_THIN_LWM2M_SERVICE)
+	else if ((cmd_type&CMD_TYPE_LWM2M_BASE) == CMD_TYPE_LWM2M_BASE) {
+		th_lwm2m_control(cmd_type, data_buf, data_len);
+	}
+#endif
+	else {
+		LOG_WRN("unknown data, dropped");
+	}
+}
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	LOG_ERR("bsdlib recoverable error: %u", err);
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	LOG_ERR("bsdlib irrecoverable error: %u", err);
+	__ASSERT_NO_MSG(false);
+}
+
+void start_execute(void)
+{
+	int err;
+
+#if !defined(CONFIG_THIN_TEST)
+	/* Initialize the inter connect library. */
+	err = inter_connect_init(data_handler);
+	if (err) {
+		LOG_ERR("Init inter_connect error: %d", err);
+		return;
+	}
+#endif /* CONFIG_THIN_TEST*/
+
+	/* AT socket creation. Handle must be valid to continue. */
+	at_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+	fds[0].fd = at_socket_fd;
+	fds[0].events = ZSOCK_POLLIN;
+	nfds += 1;
+
+	if (at_socket_fd == INVALID_DESCRIPTOR) {
+		LOG_ERR("Creating at_socket failed");
+		return;
+	}
+	LOG_INF("Start socket thread");
+	k_mutex_init(&socket_mutex);
+	k_work_init(&at_cmd_send_work, at_cmd_send);
+	k_thread_create(&socket_thread, socket_thread_stack,
+		K_THREAD_STACK_SIZEOF(socket_thread_stack),
+		at_socket_thread_fn,
+		NULL, NULL, NULL,
+		THREAD_PRIORITY, 0, K_NO_WAIT);
+	at_cmd_responded = false;
+
+	LOG_INF("The serial lte modem start with services:");
+	LOG_INF(" .AT command");
+
+#if defined(CONFIG_LTE_LINK_CONTROL)
+	LOG_INF(" .Modem control");
+#endif /* CONFIG_LTE_LINK_CONTROL */
+
+#if defined(CONFIG_THIN_MQTT_SERVICE)
+	LOG_INF(" .MQTT service");
+#endif /* CONFIG_THIN_MQTT_SERVICE */
+
+#if defined(CONFIG_THIN_LWM2M_SERVICE)
+	LOG_INF(" .LwM2M service");
+#endif /* CONFIG_THIN_LWM2M_SERVICE */
+
+#if defined(CONFIG_THIN_GPS_SERVICE)
+	LOG_INF(" .GPS service");
+	th_gps_init();
+#endif /* CONFIG_THIN_GPS_SERVICE */
+
+#if defined(CONFIG_THIN_SNTP_SERVICE)
+	LOG_INF(" .SNTP service");
+#endif /* CONFIG_THIN_SNTP_SERVICE */
+
+#if !defined(CONFIG_THIN_TEST)
+	/* Signal peer that nRF91 is ready */
+	(void)inter_connect_send(CMD_TYPE_SYNC_CMD, NULL, 0);
+#else
+	th_test_main();
+#endif	/* CONFIG_THIN_TEST */
+
+	while (true) {
+		k_sleep(K_MSEC(1000));
+		/* Put CPU to idle to save power */
+		k_cpu_idle();
```

Comment:
I think we should let the idle thread handle the CPU idling. If the main thread is not needed anymore, why don't we simply end it? There should be no harm returning here.

Reply:
I used to call MQTT alive here but removed later.
I'll remove this loop.